### PR TITLE
Give Stream::index() a concrete return type.

### PR DESCRIPTION
**Pull request by ryzhyk**
One of the DBSP API usability problems listed in #92 is that most
operators are currently generic over return type, e.g.,
`Stream::index` can return any `IndexedZSet`, forcing the programmer
to annotate every call to such operators with concrete return type,
usually `OrdIndexedZSet`.

In this commit we prototype a solution to this problem for the `index`
operator.  We split it into two: `index_generic`, which can return
any indexed Z-set type, and `index`, which specializes `index_generic`
to return `OrdIndexedZSet`.  We do the same for `index_with`.

The result is that I was able to remove all type annotations on
`index` and `index_with` calls.

We also eliminate an inefficiency in the implementation of `index_with`,
which used to create an intermediate batch instead of converting input
batch directly to output Z-set.

### CODE REVIEW

File: src/operator/index.rs [hunk 1]
```diff
@@ -114,22 +151,92 @@ where
...
+    fn index_test() {
         let root = Root::build(move |circuit| {
             let mut inputs = vec![
                 zset!{ (1, "a") => 1
-                     , (1, "a") => 2
                      , (1, "b") => 1
-                     , (1, "b") => -1
                      , (2, "a") => 1
                      , (2, "c") => 1
+                     , (1, "a") => 2
```
Comment (mbudiu-vmw) [inline src/operator/index.rs:238]: why this change?

**Review by mbudiu-vmw — APPROVED** [review]

**Review by Kixiron — APPROVED** [review]
